sigmf/archivereader.py

```diff
@@ -7,86 +7,114 @@
+                 path=None,
+                 skip_checksum=False,
+                 map_readonly=True,
+                 archive_buffer=None):
+        self.path = path
+        tar_obj = None
+        try:
+            if self.path is not None:
+                if not self.path.endswith(SIGMF_ARCHIVE_EXT):
+                    err = "archive extension != {}".format(SIGMF_ARCHIVE_EXT)
+                    raise SigMFFileError(err)
+
+                tar_obj = tarfile.open(self.path)
+
+            elif archive_buffer is not None:
+                tar_obj = tarfile.open(fileobj=archive_buffer, mode='r:')
 
-                else:
-                    print('A regular file', memb.name, 'was found but ignored in the archive')
             else:
-                print('A member of type', memb.type, 'and name', memb.name, 'was found but not handled, just FYI.')
-
-        if data_offset_size is None:
-            raise SigMFFileError('No .sigmf-data file found in archive!')
-
-        self.sigmffile = SigMFFile(metadata=json_contents)
-        valid_md = self.sigmffile.validate()
-
-        self.sigmffile.set_data_file(self.name, data_buffer=archive_buffer, skip_checksum=skip_checksum, offset=data_offset_size[0],
-                                     size_bytes=data_offset_size[1], map_readonly=map_readonly)
-
-        self.ndim = self.sigmffile.ndim
-        self.shape = self.sigmffile.shape
-
-        tar_obj.close()
+                raise ValueError('In sigmf.archivereader.__init__(), either '
+                                 '`path` or `archive_buffer` must be not None')
+
+            json_contents = None
+            data_offset_size = None
+            sigmffile_name = None
+            self.sigmffiles = []
+            data_found = False
+
+            for memb in tar_obj.getmembers():
+                if memb.isdir():  # memb.type == tarfile.DIRTYPE:
+                    # the directory structure will be reflected in the member
+                    # name
+                    continue
+
+                elif memb.isfile():  # memb.type == tarfile.REGTYPE:
+                    if memb.name.endswith(SIGMF_METADATA_EXT):
+                        json_contents = memb.name
+                        if data_offset_size is None:
+                            # consider a warnings.warn() here; the datafile
+                            # should be earlier in the archive than the
+                            # metadata, so that updating it (like, adding an
+                            # annotation) is fast.
+                            pass
+                        with tar_obj.extractfile(memb) as memb_fid:
+                            json_contents = memb_fid.read()
+
+                        sigmffile_name, _ = os.path.splitext(memb.name)
+                    elif memb.name.endswith(SIGMF_DATASET_EXT):
+                        data_offset_size = memb.offset_data, memb.size
+                        data_found = True
+                    else:
```

Comment:
It would make sense to add an additional check to see if the tar member has the ".sigmf-collection" extension. The behavior could be something like printing an info message that the file was seen, but not handled.

This behavior would be basically the same as-is, but would lay the groundwork for handling collection files in the future. It would also be nice from a user perspective for the ArchiveReader to "know" that the file is a Collection file, instead of just a "regular file"